logrus.go

```diff
@@ -0,0 +1,131 @@
+package logformatter
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/genproto/googleapis/devtools/clouderrorreporting/v1beta1"
+	"google.golang.org/genproto/googleapis/logging/type"
+)
+
+type (
+	// Interface for inspecting error objects recursively
+	// Ref: https://github.com/pkg/errors/blob/7f95ac13edff643b8ce5398b6ccab125f8a20c1a/errors.go#L30-L53
+	causer interface {
+		Cause() error
+	}
+
+	// Interface for retrieving stack frame for each error object
+	// Ref: https://github.com/pkg/errors/blob/7f95ac13edff643b8ce5398b6ccab125f8a20c1a/errors.go#L68-L73
+	stackTracer interface {
+		StackTrace() errors.StackTrace
+	}
+)
+
+// Adapt from https://github.com/googleapis/google-cloud-go/issues/1084#issuecomment-474565019
+func FormatStack(err error) (buffer []byte) {
+	if err == nil {
+		return nil
+	}
+
+	// find the inner most error with a stack
+	inner := err
+	for inner != nil {
+		if cause, ok := inner.(causer); ok {
+			inner = cause.Cause()
+			if _, ok := inner.(stackTracer); ok {
+				err = inner
+			}
+		} else {
+			break
+		}
+	}
+
+	if stackTrace, ok := err.(stackTracer); ok {
+		buf := bytes.Buffer{}
+		buf.WriteString(getGoroutineState() + "\n")
+
+		// format each frame of the stack to match runtime.Stack's format
+		var lines []string
+		for _, frame := range stackTrace.StackTrace() {
+			pc := uintptr(frame) - 1
+			fn := runtime.FuncForPC(pc)
+			if fn != nil {
+				file, line := fn.FileLine(pc)
+				lines = append(lines, fmt.Sprintf("%s()\n\t%s:%d +%#x", fn.Name(), file, line, fn.Entry()))
+			}
+		}
+		buf.WriteString(strings.Join(lines, "\n"))
+
+		buffer = buf.Bytes()
+	}
+	return
```

Comment:
What if `err` does not implement `Cause()` and `StackTrace()`?

Reply:
Nothing will be printed. This formatter is based on pkg/errors we would need to wrap the information for logging properly.

Reply:
If nothing will be printed, could we add comments to mention the input `err` should implement `Cause()` and `StackTrace()` in `example_test.go`?

Reply:
I could add the comment above FormatStack about the usage should be coupled with pkg/errors.

Reply:
done.